### PR TITLE
Game Logic: Milestone 3

### PR DESCRIPTION
**Pull request by Israel-Laguan**
# Resume

We have created the methods as described in the project requirements, added some basic tests, and *came to the conclusion that we met all of the requirements*.

# Proposal

We want to have a review of the code to know if we need to make any changes. If no changes needed we want the project to be ACCEPTED so we can move to the next milestone.

# Error-free
image

# Working logic
image

### CODE REVIEW

File: lib/tic_tac_toe.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    win_tests.each do |test|
+      next unless test[0] == test[1] && test[0] == test[2]
+
+      is_equal = true
+      mark = test[0]
+    end
+
+    [is_equal, mark]
+  end
+end
```
Comment (stratospherique) [inline lib/tic_tac_toe.rb]: Besides the issue that this method have about displaying the winner (i.e mentioned above :arrow_double_up: ), I suggest that you refactor your verify_win method as follows:

```ruby
 def check_win(player_symbol)
    # for rows
    all_equal(board[0..2], player_symbol) ||
    all_equal(board[3..5], player_symbol) ||
    all_equal(board[6..8], player_symbol)

    # and analogical as above for columns and diagonals
  end

  def all_equal(subarray, player_symbol)
    return true if subarray.uniq.size == 1 && subarray.uniq == player_symbol
  end

```

File: bin/main.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  msg(message)
+  result= gets.chomp
+  result
+end
+
+def show_board(board)
+  puts " #{board[0] == 10 ? 'X' : board[0]} | #{board[1] == 10 ? 'X' : board[1]} | #{board[2] == 10 ? 'X' : board[2]}"
+  puts " #{board[3] == 10 ? 'X' : board[3]} | #{board[4] == 10 ? 'X' : board[4]} | #{board[5] == 10 ? 'X' : board[5]}"
+  puts " #{board[6] == 10 ? 'X' : board[6]} | #{board[7] == 10 ? 'X' : board[7]} | #{board[8] == 10 ? 'X' : board[8]}"
+end
```
Comment (stratospherique) [inline bin/main.rb]: - Objects in the game should communicate explicitly through messages(methods) not implicitly through global methods. You can associate these methods with an existing class, or create a sophisticated one for them. :nerd_face:

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive work so far :clap: . However, some requirements needs to be implemented in order for us to approve your project. check below :arrow_down: 

#### Implementation details 

* [x] All classes are kept in `lib` directory.
  - TicTacToe module and its classes.
- Check the code comments :arrow_down: for more suggestions.

#### Cases where the game is not showing the right output:

```
Player alex is your turn
Choose your hand
9
 X | 0 | 0
 4 | X | 6
 7 | 8 | X
Player 007 is your turn
Choose your hand
```
  - No one is victorious in this case, at least not immediately (i.e next turn)

### That's it :relaxed: When finished, kindly resubmit another code review request :pray: 

**Ahmed Mahfoudh**

**Comment by Israel-Laguan** [comment]
We addresed your comments, thanks for reviewing:
>     * [ ]  All classes are kept in `lib` directory.
>       
>       * TicTacToe module and its classes.

**Solved!**

> #### Cases where the game is not showing the right output:
> 
> ```
> Player alex is your turn
> Choose your hand
> 9
>  X | 0 | 0
>  4 | X | 6
>  7 | 8 | X
> Player 007 is your turn
> Choose your hand
> ```
> 
>     * No one is victorious in this case, at least not immediately (i.e next turn)

Turns out the problem was that we started checking for winner on turn 6 instead of turn 5, **solved**

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## CHANGES REQUIRED

### Hi team :wave: 
Nice, implementing the changes requested by colleague, but their a still some blindspots :slightly_smiling_face: 

There are improvements we need to make.

- [x] The application should tell if there is a draw 

- [x] The application should inform the players if a move is a win move.
 
Currently, your application does notify the players the status of their game, rather it exits when there is a win or draw. Preferrable, it should ask the players if they would play again, with a yes or no option to choose from.

- [**OPTIONAL**] In your project instead of having a module, I want to suggest you create classes, at least two classes from: Game, Board or Player. 

- [x] Use attr_readers/attr_writers instead of atrr_acessors; attr_acessor, solves the problem but not always the best choice. For learning purposes, we would like to see you implement attr_reader and attr_writer in this project, this would help consolidate your understanding of the concept.

### After implementing these changes submit another form for a code review.

#### ``` Have a great day ```

```Uchechukwu Henry Inyama```

**Comment by Israel-Laguan** [comment]
# Changes

We made the changes as recommended by @uche-inyama 
>     * [x]  The application should tell if there is a draw
> 
>     * [x]  The application should inform the players if a move is a win move.
> 
>     * [x]  Use attr_readers/attr_writers instead of atrr_acessors; attr_acessor, solves the problem but not always the best choice. For learning purposes, we would like to see you implement attr_reader and attr_writer in this project, this would help consolidate your understanding of the concept.

**Review by AdnanAfsari — APPROVED** [review]
**Status: Approved**

Hello @Israel-Laguan @izaiasneto4 ✋ 

Nice improvement, looks good now so its a 👍 from me.

Happy Coding
Adnan Afsari

**Pull request merged**